jax/numpy/lax_numpy.py

```diff
@@ -979,6 +979,32 @@ def nan_reduction(a, axis=None, out=None, keepdims=False, **kwargs):
 nansum = _make_nan_reduction(onp.nansum, sum, 0, nan_if_all_nan=False)
 nanprod = _make_nan_reduction(onp.nanprod, prod, 1, nan_if_all_nan=False)
 
+@_wraps(onp.nanmean)
+def nanmean(a, axis=None, dtype=None, out=None, keepdims=False):
+  if out is not None:
+    raise ValueError("nanmean does not support the `out` argument.")
+  # Check and Count all NaN's per axis
+  nan_mask = isnan(a)
+  nan_cnt = sum(nan_mask, axis=axis, dtype=bool_, keepdims=keepdims)
```

Comment:
I don't think the logic here is quite correct. Here, you're performing a bool sum reduction to test whether there exists a nan anywhere in the input array. `dtype=bool_` means you get a bool result, i.e., whether there is a nan present or not. If not, you do a regular `mean`. If so, you make a correction to the normalizer. However, that correction is looks like it is either 1 or 0, which won't handle the case where there is more than 1 nan in the input array.

I think we can simplify this a little bit.

I would have guessed that the right algorithm here would be to:
a) count the number of non-nan values — you can do this something like:
```
normalizer = sum(logical_not(isnan(a)), axis=axis, dtype=int32, keepdims=keepdims)
```

b) divide the sum of the non-nan values by the non-nan count?
```
normalizer = lax.convert_element_type(normalizer, dtype)
output = lax.div(nansum(a, axis=axis, dtype=dtype, keepdims=keepdims), normalizer)
```
?

(That's may not be precisely right, especially with respect to type promotion but you get the general idea.)

Reply:
Yeah, that algorithm looks right! I'll throw together this with some tests now.